formats/json/commonMain/src/kotlinx/serialization/json/internal/JsonTreeReader.kt

```diff
@@ -75,9 +83,28 @@ internal class JsonTreeReader(
         return when (lexer.peekNextToken()) {
             TC_STRING -> readValue(isString = true)
             TC_OTHER -> readValue(isString = false)
-            TC_BEGIN_OBJ -> readObject()
+            TC_BEGIN_OBJ -> {
+                // 500 is an arbitrary constant
+                val result = if (++stackDepth > 500) {
```

Comment:
It seems that it should be checking level for equality, e.g. (`++stackDepth == RECUSIVE_LIMIT`). In the current implemented a new `DeepRecursiveFunction` invocation context will be created on each recursion level beyond the limit.

Reply:
Not really because once we're in `readDeepRecursive`, we no longer can re-enter this function (moreover, when nested objects are separated via `JsonArray`, equality check will prevent us from maintaining multiple deep-recursive scopes)